### ISSUE TITLE
Possible not server resource intensive solution for Eve placement

### ISSUE THREAD

**Issue by Arcurus** [jasonrohrer/OneLife#624]
**Problem:**
1. Currently eves spawn quite close to each other, which leads to being spawned in dry out areas with nearly no natural food.

2. Since Eve spawn moves from East to West, but never back developing a road system is economic not reasonable, since towns stay too short in the same area so that the benefit of developing a road system is too less compared to the cost / time to build it.

Simply increasing the spawn distance would lead to increase the second problem, while reducing the spawn distance would increase the first problem. Therefore the following suggest to use another Eve spawn algorithm based on the last average player positions as Eve spawn center to solve the first issue and a calculated score of the possible Eve spawn locations based on players nearby, not dry natural springs and availability of wild food. To limit the needed server resources only Y semi random possible Eve spawn locations are chosen and a score of each calculated. The highest scored Eve spawn location is then chosen. If the score is too low, the semi random chosen max /min distance from the player location based eve spawn center is increased and another Y semi random possible Eve spawn locations are calculated. This is repeated again until an high enough score is found, or until Z max steps to limit server resource usage.

Past Eve spawn locations:
https://onemap.wondible.com

**Outlined Algorithm: (the chosen numbers are just a hint, for sure you know better the numbers then me)**

1. Calculate the Eve spawn center from the average last living 100 players positions.

2. Calculate 10 random eve spawn locations in an max distance of 300 tiles* from the eve spawn center (*In each iteration the max distance is increased by 100 tiles see 5.) 

3. Calculate the score of each chosen Eve spawn location looking at a distance of 100 tiles with: 
+10 for each not dry well site
-5 for each dry well site
+1 for each Natural food resource
-1 for each empty Natural food resource
-10 for each player in this area

4. Choose the best scored location out of the 10 calculated Eve spawn locations

5. Do step 2 to 5 again until at least a location with score >20 is found, while increasing the max and min distance from the calculated eve spawn center by 100 tiles with each iteration (optional limit at max Z iterations to limit possible resource usage).

Since the number of iterations is limited the suggested Eve spawn algorithm would have an constant max cost.

**Comment by laccy** [comment]
yes, it should have some food around, natural or not natural. 
it's a problem you can't solve, you get into an old city with high hopes and no food left, it's worse than the wilderness. No reason to spawn in an Eve when there is 0 food anywhere +-1 or 2 well sites in any direction.

also, why remove the last berry from bushes? I thought that's a feature. That allows a few people to move the distance, it's not overpowered to have 1 berry on bushes. late gen is 1 pip, stopping there is more of a waste of time, you lose a pip for getting there, you gain one for one berry.

For Eves is 5 points so you can stay alive and find a zone with food.

**Comment by jasonrohrer** [comment]
The problem is with Step 1 above.

If players are spread out (which they are), the center (average location of living players) will be between villages, likely in a stripped-bare zone.

Then you go out 300 tiles from there, and what do you find?  Villages.

So, there's no center or radius-based thing that will work.  Too many bad edge cases.

The current idea is to place Eve on the western "horizon" of civilization.  Apparently, that's not working very well.  But that's a good place to look for untouched wilderness.

**Comment by jasonrohrer** [comment]
What about just looking to the west of the furthest-west current homeland?

**Comment by TwistedHBG** [comment]
> 
> 
> What about just looking to the west of the furthest-west current homeland?

I hop on a horse, run 8k tiles west, make a well. Bam, towns are now 8k tiles apart.

I feel like one of the main reasons Eves spawn in populated areas is because the Eve jump size is currently set to 50. It's very easy to get ahead of the Eve curve and create towns in locations where Eves will only spawn in a few days.

How many Eves spawn per day currently?

What if the Eve jump size was increased to 100 (lower than the old 200) and the Eve belt height was doubled from 5 to 10. This would keep westward drift the same as it is now but a bit more spread out.

With 25 daily Eves this would mean daily westward drift would be under still be about 250 like it is now, which I think is fine. Maximum N/S distance between towns would be 1k, which is decently far but more than manageable.

**Comment by jasonrohrer** [comment]
Twisted, I would only be counting active homelands (where bb are born).

So you could hop on a horse and make a homeland out there, but unless you maintained it, we would resume placing future Eves closer after you died.

Also, we don't really know what people are going to do.... how fast are they going to spread out?

The current fixed march of Eve doesn't really account for this variability.

As you've pointed out, players are moving West on their own faster than Eve placement moves West, so there is overlap.

Likewise, if players stopped spreading to the west on their own, we might have a situation where Eves walk too fast to the West and get farther away than they should be.

It does make sense to figure out where players actually are, and put Eve to the west of that.

**Comment by jasonrohrer** [comment]
Oh, yeah... if you placed a temp homeland 8K to the west, and some Eve spawned out there (very lucky timing), then that Eve would settle out there, and if they survived, they would push the spawn of future Eves further west forever.

But how long would an Eve that far away survive?  They won't easily be able to trade, right?

I guess the main thing is walking far enough west that we have some non-dry well sites around... hmm...

**Comment by Arcurus** [comment]
>The problem is with Step 1 above.
If players are spread out (which they are), the center (average location of living players) will be between villages, likely in a stripped-bare zone.
Then you go out 300 tiles from there, and what do you find? Villages.
So, there's no center or radius-based thing that will work. Too many bad edge cases.

Point 5 solves this problem. The iteration is repeated with increased radius until an good enough location is found. One iteration should not be that costly, so many iterations could be done, so above outlined algorithm should always find a quite good spot with reasonable resource usage.

In short is starts to search in the center, but if no good location is found it increases the search radius. If needed i can outline the alg in pseudo c++ code.

The outlined algorithm also solves all the edge cases with eves moving here and there, it always starts the search in the middle of the current players, so there is high chance that a close untouched land is found through the algorithm.

So it solves many problems at once. Eves will have an "untouched" spawn but still will be quite close to current villages, also there is no artificial eat west movement, so road systems can be reused.

**Comment by Arcurus** [comment]
Here the pseudo c++ code, sorry that the code formatting wont work, dont know why. Also it removed my spaces, just edit it then you get them.

`

void getEvePosition(int *outX, int *outY) {
	int averagePlayerPositions = 100;
	int radiuslncrement = 100;
	int maxIterations = 100;
	int iterations = 0
	int minScore = 20;
	int searchRadius = 300;
	int centerX;
	int centerY;
	int score;

	calculateAverageLastPlayerPosition(averagePlayerPositions, &centerX, &centerY);

	do {
		score = CalculateBestScoredPlace(searchRadius, centerX, centerY, outX, outY);
		searchRadius += radiuslncrement;
		iterations++;
	} while (score < minScore  && iterations <= maxIterations)  )
}

CalculateBestScoredPlace(int searchRadius, int centerX, int centerY, int *outX, int *outY) {
	int x;
	int y;
	int score = -10000000000;

	for(i = 0, i < 10, i++) {
		calculateRandomPointFomCenter(searchRadius, centerX, centerY, x, y);
		int tmpScore = CalculateLocationScore(x, y);
		if (tmpScore > score) {
			score = tmpScore;
			*outX = x;
			*outY = y;
		}
	}
	return score;
}
`

**Comment by jasonrohrer** [comment]
Yeah, the increasing radius is the part that gets expensive...  I'm really looking for something simpler and more sure-fire than this.

**Comment by Arcurus** [comment]
even if the radius increases, the score is only calculated local / the next X tiles around the randomly chosen location, so the effort for each calculation stays the same. Should not be much expensive then quer-ring the location data for a player.

**Comment by Arcurus** [comment]
placing Eve spawn West of existing town will solve problem 1, that there is no food / spring but will increase Problem 2, that Eve spawn will move much faster to the West and therefore give not much time for the villages to connect.
